File: 2024/06/main.py
#!/usr/bin/env python3

import sys
import tqdm

def getStart(data):
    R = len(data)
    C = len(data[0])
    for r in range(R):
        for c in range(C):
            if data[r][c] in ["^",">","v","<"]:
                return r,c

def doPart1():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    data = []
    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue
            data.append(list(line))

    sr,sc = getStart(data)
    data[sr][sc] = "X"

    R = len(data)
    C = len(data[0])
    numTraveled = 1
    DIR = ((-1,0),(0,1),(1,0),(0,-1))
    curDir = 0
    while True:
        if data[sr][sc] != "X":
            data[sr][sc] = "X"
            numTraveled += 1
        # Look ahead one space
        nr = sr + DIR[curDir][0]
        nc = sc + DIR[curDir][1]
        # Check if this has left the board
        if nr < 0 or nr == R or nc < 0 or nc == C:
            break
        nextSpace = data[nr][nc]
        if nextSpace == "#":
            curDir = (curDir + 1) % 4
        sr += DIR[curDir][0]
        sc += DIR[curDir][1]

    print(numTraveled)

# Combine part 1 and part 2 together as we need part 1 to get the set for part 2
def doPart2Again():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    data = []
    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue
            data.append(line)

    sr,sc = getStart(data)

    numLoops = 0
    # Just try adding an obstacle in all possible locations???
    R = len(data)
    C = len(data[0])
    DIR = ((-1,0),(0,1),(1,0),(0,-1))

    cr = sr
    cc = sc
    curDir = 0

    # Lets use sets this time so I don't have to update the grid...
    PATH_RC = set()
    LOOP_RC = set()

    # "kinda" same as last time
    while True:
        if (cr,cc,curDir) in LOOP_RC:
            numLoops += 1
            break
        LOOP_RC.add((cr,cc,curDir))
        PATH_RC.add((cr,cc))

        nr = cr+DIR[curDir][0]
        nc = cc+DIR[curDir][1]

        if not (0 <= nr < R and 0 <= nc < C):
            break
        if data[nr][nc] == "#":
            curDir = (curDir+1)%4
        else:
            cr = nr
            cc = nc
    print(len(PATH_RC))

    for prc in tqdm.tqdm(PATH_RC):
        ob_r = prc[0]
        ob_c = prc[1]

        cr = sr
        cc = sc
        curDir = 0

        # Lets use sets this time so I don't have to update the grid...
        PATH_RC = set()
        LOOP_RC = set()

        # "kinda" same as last time
        while True:
            if (cr,cc,curDir) in LOOP_RC:
                numLoops += 1
                break
            LOOP_RC.add((cr,cc,curDir))
            PATH_RC.add((cr,cc))

            nr = cr+DIR[curDir][0]
            nc = cc+DIR[curDir][1]

            if not (0 <= nr < R and 0 <= nc < C):
                if data[ob_r][ob_c]=="#":
                    numPath = len(PATH_RC) # We can still get the part one answer...
                break
            if data[nr][nc] == "#" or (nr == ob_r and nc == ob_c):
                curDir = (curDir+1)%4
            else:
                cr = nr
                cc = nc
    print(numLoops)

def doPart2():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    data = []
    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue
            data.append(line)

    sr,sc = getStart(data)

    numLoops = 0
    # Just try adding an obstacle in all possible locations???
    R = len(data)
    C = len(data[0])
    DIR = ((-1,0),(0,1),(1,0),(0,-1))
    # Should just solve one maze without an ob and use the set of possible guard positions as the input instead...
    for ob_r in tqdm.tqdm(range(R)):
        for ob_c in range(C):
            cr = sr
            cc = sc
            curDir = 0

            # Lets use sets this time so I don't have to update the grid...
            PATH_RC = set()
            LOOP_RC = set()

            # "kinda" same as last time
            while True:
                if (cr,cc,curDir) in LOOP_RC:
                    numLoops += 1
                    break
                LOOP_RC.add((cr,cc,curDir))
                PATH_RC.add((cr,cc))

                nr = cr+DIR[curDir][0]
                nc = cc+DIR[curDir][1]

                if not (0 <= nr < R and 0 <= nc < C):
                    if data[ob_r][ob_c]=="#":
                        numPath = len(PATH_RC) # We can still get the part one answer...
                    break
                if data[nr][nc] == "#" or (nr == ob_r and nc == ob_c):
                    curDir = (curDir+1)%4
                else:
                    cr = nr
                    cc = nc
    print(numPath)
    print(numLoops)


def main():
    # doPart1()
    # doPart2()
    doPart2Again()

if __name__ == "__main__":
    main()

File: 2024/06/main.cpp

#include "../helpers/extras.hh"
#include <array>
#include <set>

std::array<int,2> getStart(std::vector<std::string>& data)
{
    const int R = data.size();
    const int C = data[0].length();
    for (int r = 0; r < R; r++) {
        for (int c = 0; c < C; c++) {
            if (data[r].c_str()[c] == '^') return std::array<int,2>{r,c};
        }
    }
    return std::array<int,2>{-1,-1};
}

std::set<std::array<int,2>> doPart1(const char* filename) // 4665
{
    std::ifstream file(filename);
    std::string line;
    std::vector<std::string> data;
    while (getline(file, line)) {
        stripString(line);
        data.push_back(line);
    }
    file.close();

    const int R = data.size();
    const int C = data[0].length();

    std::array<int,2> start_pos = getStart(data);
    const int sr = start_pos[0];
    const int sc = start_pos[1];

    std::set<std::array<int,2>> SEEN;

    std::array<std::array<int,2>,4> DIR; // = {{-1,0},{0,1},{1,0},{0,-1}};
    DIR[0] = {-1, 0};
    DIR[1] = { 0, 1};
    DIR[2] = { 1, 0};
    DIR[3] = { 0,-1};
    int curDir = 0;
    int cr = sr;
    int cc = sc;

    while (true) {
        SEEN.insert(std::array<int,2>{cr,cc});
        const int nr = cr+DIR[curDir][0];
        const int nc = cc+DIR[curDir][1];

        if (! (0 <= nr && nr < R && 0 <= nc && nc < C)) {
            break;
        }
        if (data[nr][nc] == '#') {
            curDir = (curDir+1)%4;
        } else {
            cr = nr;
            cc = nc;
        }
    }
    printf("%ld\n", SEEN.size());
    return SEEN;
}

void doPart2(const char* filename, std::set<std::array<int,2>>& SEEN) // 1688
{
    std::ifstream file(filename);
    std::string line;
    std::vector<std::string> data;
    while (getline(file, line)) {
        stripString(line);
        data.push_back(line);
    }
    file.close();

    const int R = data.size();
    const int C = data[0].length();

    std::array<std::array<int,2>,4> DIR;
    DIR[0] = {-1, 0};
    DIR[1] = { 0, 1};
    DIR[2] = { 1, 0};
    DIR[3] = { 0,-1};

    std::array<int,2> start_pos = getStart(data);
    const int sr = start_pos[0];
    const int sc = start_pos[1];

    int numLoops = 0;
    for (auto prc : SEEN) {
        const int ob_r = prc[0];
        const int ob_c = prc[1];
        int cr = sr;
        int cc = sc;
        int curDir = 0;

        std::set<std::array<int,3>> LOOP_RC;

        while (true) {
            if (LOOP_RC.find(std::array<int,3>{cr,cc,curDir})!=LOOP_RC.end()) {
                numLoops +=1;
                break;
            }
            LOOP_RC.insert(std::array<int,3>{cr,cc,curDir});
            const int nr = cr + DIR[curDir][0];
            const int nc = cc + DIR[curDir][1];

            if (! (0 <= nr && nr < R && 0 <= nc && nc < C)) {
                break;
            }
            if (data[nr][nc] == '#' || (nr == ob_r && nc == ob_c)) {
                curDir = (curDir+1)%4;
            } else {
                cr = nr;
                cc = nc;
            }
        }
    }
    printf("%d\n", numLoops);
}

int main(int narg, char* args[])
{
    std::string filename = "example.dat";
    if (narg > 1) filename = std::string(args[1]);
    std::set<std::array<int,2>> SEEN = doPart1(filename.c_str());
    doPart2(filename.c_str(), SEEN);
    return 0;
}
